### PR TITLE
refactor IPT plugin to use libhook

### PR DESCRIPTION
**Pull request by BonusPlay**

### CODE REVIEW

**Comment by BonusPlay** [comment]
Should be ready. Migrated to libhook for cleaner plugin exit. Additionally added new trap type to libhook.

Sidenote: I'll try to cleanup plugins_ex.h and codemon in upcoming PRs.

**Comment by tklengyel** [comment]
@drakvuf-jenkins Test this please

File: src/libhook/hooks/catchall.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
+/*********************IMPORTANT DRAKVUF LICENSE TERMS***********************
+ *                                                                         *
+ * DRAKVUF (C) 2014-2022 Tamas K Lengyel.                                  *
```
Comment (tklengyel) [inline src/libhook/hooks/catchall.cpp:3]: 2023

File: src/libhook/hooks/catchall.hpp [hunk 2]
```diff
@@ -0,0 +1,201 @@
+/*********************IMPORTANT DRAKVUF LICENSE TERMS***********************
+ *                                                                         *
+ * DRAKVUF (C) 2014-2022 Tamas K Lengyel.                                  *
```
Comment (tklengyel) [inline src/libhook/hooks/catchall.hpp:3]: 2023

**Comment by BonusPlay** [comment]
Fixed.

**Comment by tklengyel** [comment]
@drakvuf-jenkins Test this please

**Comment by tklengyel** [comment]
FYI
```
** CID 404291:  Memory - illegal accesses  (UNINIT)


________________________________________________________________________________________________________
*** CID 404291:  Memory - illegal accesses  (UNINIT)
/src/libhook/hooks/catchall.cpp: 137 in libhook::CatchAllHook::CatchAllHook(libhook::CatchAllHook&&)()
131         }
132     }
133     
134     CatchAllHook::CatchAllHook(CatchAllHook&& rhs) noexcept
135         : BaseHook(std::forward<BaseHook>(rhs))
136     {
>>>     CID 404291:  Memory - illegal accesses  (UNINIT)
>>>     Using uninitialized value "this->trap_" when calling "swap".
137         std::swap(this->trap_, rhs.trap_);
138         std::swap(this->callback_, rhs.callback_);
139     }
140     
141     CatchAllHook& CatchAllHook::operator=(CatchAllHook&& rhs) noexcept
142     {

** CID 404290:  Memory - illegal accesses  (OVERRUN)


________________________________________________________________________________________________________
*** CID 404290:  Memory - illegal accesses  (OVERRUN)
/src/plugins/plugins_ex.h: 741 in pluginex::createCr3Hook<PluginResult, unsigned int (ipt::*)(drakvuf *, drakvuf_trap_info *)>(T2, int)::[lambda(T1 &&...) (instance 1)]::operator ()<drakvuf *, drakvuf_trap_info *>(unsigned int, T1 &&...) const()
735     template<typename Params, typename Callback>
736     std::unique_ptr<libhook::Cr3Hook> pluginex::createCr3Hook(Callback cb, int ttl)
737     {
738         static_assert(std::is_base_of_v<PluginResult, Params>, "Params must derive from PluginResult");
739         auto hook = libhook::Cr3Hook::create<Params>(this->drakvuf, [=](auto&& ...args) -> event_response_t
740         {
>>>     CID 404290:  Memory - illegal accesses  (OVERRUN)
>>>     Overrunning array of 16 bytes at byte offset 23 by dereferencing pointer "this->cb".
741             return std::invoke(cb, static_cast<typename class_type<Callback>::type*>(this), args...);
742         }, ttl);
743         if (hook)
744             static_cast<Params*>(hook->trap_->data)->plugin_ = this;
745         else
746             PRINT_DEBUG("[WARNING] libhook failed to setup a trap, returning nullptr!\n");

** CID 404289:  Uninitialized members  (UNINIT_CTOR)
/src/libhook/hooks/catchall.cpp: 151 in libhook::CatchAllHook::CatchAllHook(drakvuf *, std::function<unsigned int (drakvuf *, drakvuf_trap_info *)>)()


________________________________________________________________________________________________________
*** CID 404289:  Uninitialized members  (UNINIT_CTOR)
/src/libhook/hooks/catchall.cpp: 151 in libhook::CatchAllHook::CatchAllHook(drakvuf *, std::function<unsigned int (drakvuf *, drakvuf_trap_info *)>)()
145         return *this;
146     }
147     
148     CatchAllHook::CatchAllHook(drakvuf_t drakvuf, cb_wrapper_t cb)
149         : BaseHook(drakvuf),
150           callback_(cb)
>>>     CID 404289:  Uninitialized members  (UNINIT_CTOR)
>>>     Non-static class member "trap_" is not initialized in this constructor nor in any functions that it calls.
151     {}
152     

** CID 404288:  Memory - illegal accesses  (WRAPPER_ESCAPE)
/src/libhook/hooks/cr3.hpp: 185 in libhook::Cr3Hook::create<PluginResult>(drakvuf *, std::function<unsigned int (drakvuf *, drakvuf_trap_info *)>, int)()


________________________________________________________________________________________________________
*** CID 404288:  Memory - illegal accesses  (WRAPPER_ESCAPE)
/src/libhook/hooks/cr3.hpp: 185 in libhook::Cr3Hook::create<PluginResult>(drakvuf *, std::function<unsigned int (drakvuf *, drakvuf_trap_info *)>, int)()
179     
180         static_assert(std::is_base_of_v<CallResult, Params>, "Params must derive from CallResult");
181         static_assert(std::is_default_constructible_v<Params>, "Params must be default constructible");
182     
183         // populate backref
184         auto* params = new Params();
>>>     CID 404288:  Memory - illegal accesses  (WRAPPER_ESCAPE)
>>>     The internal representation of local "hook" escapes into "params->hook_", but is destroyed when it exits scope.
185         params->hook_ = hook.get();
186         hook->trap_->data = static_cast<void*>(params);
187     
188         if (!drakvuf_add_trap(drakvuf, hook->trap_))
189         {
190             PRINT_DEBUG("[LIBHOOK] failed to create cr3 trap!\n");

** CID 404287:  Memory - illegal accesses  (OVERRUN)


________________________________________________________________________________________________________
*** CID 404287:  Memory - illegal accesses  (OVERRUN)
/src/plugins/plugins_ex.h: 819 in pluginex::createCatchAllHook<PluginResult, unsigned int (ipt::*)(drakvuf *, drakvuf_trap_info *)>(T2, int)::[lambda(T1 &&...) (instance 1)]::operator ()<drakvuf *, drakvuf_trap_info *>(unsigned int, T1 &&...) const()
813     template<typename Params, typename Callback>
814     std::unique_ptr<libhook::CatchAllHook> pluginex::createCatchAllHook(Callback cb, int ttl)
815     {
816         static_assert(std::is_base_of_v<PluginResult, Params>, "Params must derive from PluginResult");
817         auto hook = libhook::CatchAllHook::create<Params>(this->drakvuf, [=](auto&& ...args) -> event_response_t
818         {
>>>     CID 404287:  Memory - illegal accesses  (OVERRUN)
>>>     Overrunning array of 16 bytes at byte offset 23 by dereferencing pointer "this->cb".
819             return std::invoke(cb, static_cast<typename class_type<Callback>::type*>(this), args...);
820         }, ttl);
821         if (hook)
822             static_cast<Params*>(hook->trap_->data)->plugin_ = this;
823         else
824             PRINT_DEBUG("[WARNING] libhook failed to setup a trap, returning nullptr!\n");

** CID 404286:  Memory - illegal accesses  (WRAPPER_ESCAPE)
/src/libhook/hooks/catchall.hpp: 184 in libhook::CatchAllHook::create<PluginResult>(drakvuf *, std::function<unsigned int (drakvuf *, drakvuf_trap_info *)>, int)()


________________________________________________________________________________________________________
*** CID 404286:  Memory - illegal accesses  (WRAPPER_ESCAPE)
/src/libhook/hooks/catchall.hpp: 184 in libhook::CatchAllHook::create<PluginResult>(drakvuf *, std::function<unsigned int (drakvuf *, drakvuf_trap_info *)>, int)()
178     
179         static_assert(std::is_base_of_v<CallResult, Params>, "Params must derive from CallResult");
180         static_assert(std::is_default_constructible_v<Params>, "Params must be default constructible");
181     
182         // populate backref
183         auto* params = new Params();
>>>     CID 404286:  Memory - illegal accesses  (WRAPPER_ESCAPE)
>>>     The internal representation of local "hook" escapes into "params->hook_", but is destroyed when it exits scope.
184         params->hook_ = hook.get();
185         hook->trap_->data = static_cast<void*>(params);
186     
187         if (!drakvuf_add_trap(drakvuf, hook->trap_))
188         {
189             PRINT_DEBUG("[LIBHOOK] failed to create CatchAll trap!\n");

**Comment by BonusPlay** [comment]
@tklengyel would you mind sharing how you got those?

`>>>     CID 404286:  Memory - illegal accesses  (WRAPPER_ESCAPE)` it's correct, but accounted in other places. The diagnostic is overly happy here.

`*** CID 404291:  Memory - illegal accesses  (UNINIT)` is technically correct, I'll fix the UB in another PR.

`*** CID 404287:  Memory - illegal accesses  (OVERRUN)` is what worries me. I've requested access to the project on coverity.

**Pull request merged**